### PR TITLE
[WebProfilerBundle] Show an error message if the <body> tag is not closed

### PR DESCRIPTION
**Pull request by Deuchnord**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | _none_   
| License       | MIT
| Doc PR        | _none_

In some situations, it can happen that the HTML `<body>` tag is not closed, which leads the Web Debug Toolbar to disappear without any explanation, which is quite troubling for inexperienced developers.
In this PR, I propose to show a clear error inviting the developer to fix their DOM.

A bar at the bottom displaying "An error occurred while loading the debug toolbar. Did you forget to close the <body> tag?"

### CODE REVIEW

**Comment by thewilkybarkid** [comment]
Reading the comment in the existing code:

> The WDT is only injected on well-formed HTML (with a proper &lt;/body> tag).

It’s [valid to not close (or even start) the `<body>`](https://html.spec.whatwg.org/multipage/sections.html#the-body-element) (not that I’d do it myself). Same applies to `<html>`.

So could try (in order):

1. Before `</body>` if it exists.
2. Before `</html>` if it exists.
3. At the end.

if the HTML5 doctype is present.

**Comment by Deuchnord** [comment]
Actually, according to [W3C's specifications](https://www.w3.org/TR/2011/WD-html5-author-20110809/syntax.html#syntax-tag-omission), a bunch of element end tags (and even start tags) may be omitted, but only on certain circumstances:

> A `body` element's start tag may be omitted if the element is empty, or if the first thing inside the body element is not a space character or a comment, except if the first thing inside the body element is a script or style element.
>
> A `body` element's end tag may be omitted if the body element is not immediately followed by a comment.

Given all these constraints, I'm not sure that's a good practice to omit those end tags, as such omissions can lead to some erratic behavior, ergo I don't think the WDT should provide support for this kind of syntax…

Furthermore, this flexibility seems to be specific to HTML (whatever the version of the specification you use). You cannot do it on XHTML, because it's based on XML, which is more strict. Adding such a verification based on the DocType looks to me like adding too much complexity to address an issue that would be not considered as a good practice anyway.

**Comment by Deuchnord** [comment]
A small conversation on Twitter enlightened that [Google actually recommends to not close the tags](https://google.github.io/styleguide/htmlcssguide.html#Optional_Tags), and even to omit the ones that are optional. Even though this recommendation looks weird to me, the fact that they recommend this means that many developers will eventually follow it, so I'm finally changing he behavior to show the WDT in these cases.

**Comment by dunglas** [comment]
Regarding optional HTML elements and Google best practices, see also https://github.com/symfony/recipes/pull/422

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 1]
```diff
@@ -131,6 +131,20 @@ protected function injectToolbar(Response $response, Request $request, array $no
                     'csp_style_nonce' => isset($nonces['csp_style_nonce']) ? $nonces['csp_style_nonce'] : null,
                 ]
             ))."\n";
+
+            $content = substr($content, 0, $pos).$toolbar.substr($content, $pos);
+            $response->setContent($content);
+        } elseif (false !== $pos = strripos($content, '<body>')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: it's pretty common to add attributes to the opening tag

File: src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php [hunk 2]
```diff
@@ -131,6 +131,20 @@ protected function injectToolbar(Response $response, Request $request, array $no
...
+            // We want the debug bar to be after the <body> tag, not before.
+            $pos += 6;
+
+            $toolbar = "\n".str_replace("\n", '', $this->twig->render(
+                '@WebProfiler/Profiler/toolbar_error.html.twig',
+                [
+                    'error_message' => 'An error occurred while loading the debug toolbar. Did you forget to close the <body> tag?',
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/WebProfilerBundle/EventListener/WebDebugToolbarListener.php]: which error? that's not an error, is it?

**Review by nicolas-grekas — COMMENTED** [review]
not convinced this is useful... :)

**Comment by fabpot** [comment]
Wouldn't that break when an HTML fragment is sent for instance (where we don't want to add the WDT)?

**Comment by nicolas-grekas** [comment]
> Wouldn't that break when an HTML fragment is sent for instance

it would! :-1: here

**Comment by fabpot** [comment]
Closing then.

**Pull request closed without merging**